README.md
⚡ Lightweight Node.js HTTP client.

![logo](./logo.png)

## Install

```
npm i lightning-request
```

## Why lightning-request?

lightning-request is the most lightweight HTTP client for Node, it provides a number of extremely useful features.

## Use lightning-request

First, require the library.

```js
const { request } = require('lightning-request');
```

Then let's make a request in an async function.

```
(async function() {
  try {
    const result = await request({
      url: 'https://github.com/node-labx/lightning-request',
    });
    console.log(result.statusCode); // response status code
    console.log(result.data); // response data
  } catch (error) {
    console.log(error);
  }
})();
```

## Request Config

These are the available config options for making requests. Only the url is required. Requests will default to GET if method is not specified.

```
{
  // `url` is the server URL that will be used for the request
  url: 'http://www.example.com/',

  // `method` is the request method to be used when making the request
  method: 'get', // default

  // `headers` are custom headers to be sent
  headers: {'Content-Type': 'application/json'},

  // `data` is the data to be sent as the request body
  data: {
    foo: 'bar'
  },

  // `timeout` specifies the number of milliseconds before the request times out.
  // If the request takes longer than `timeout`, the request will be aborted.
  timeout: 1000, // default is `15000` milliseconds

  // `responseType` indicates the type of data that the server will respond with
  // options are: 'json', 'text', 'buffer'，'stream'
  responseType: 'json', // default

  // `agent` define a custom agent to be used when performing http or https requests,
  // respectively, in node.js. This allows options to be added like `keepAlive` that are not enabled by default.
  agent: new http.Agent({ keepAlive: true }),
}
```

## Response Schema

The response for a request contains the following information.

```
{
  // `body` is the raw buffer data from the server response
  body: <Buffer >

  // `statusCode` is the HTTP status code from the server response
  statusCode: 200,

  // `statusMessage` is the HTTP status message from the server response
  statusMessage: 'OK',

  // `headers` the headers that the server responded with All header names are lower cased
  headers: {},

  // `data` is the response data that was provided by the server
  data: {}
}
```

## Contributing

- Fork this repo
- Clone your repo
- Install dependencies
- Checkout a feature branch
- Feel free to add your features
- Make sure your features are fully tested
- Open a pull request, and enjoy <3

## License

[MIT](LICENSE)


CHANGELOG.md
## v1.0.0

2021-02-07

- module.exports 导出修改，原默认直接导出 request 方法，现改成一个对象，包含两个方法 request / syncRequest
- 新增测试 mock server
- 修改测试用例


lib/request.js
const http = require('http');
const https = require('https');
const { URL } = require('url');
const qs = require('querystring');
const pump = require('pump');
const Response = require('./response');
const createError = require('./createError');
const enhanceError = require('./enhanceError');
const pkg = require('../package.json');

/**
 * @param {String} options.url `url` is the server URL that will be used for the request
 * @param {String} options.method `method` is the request method to be used when making the request
 * @param {Object} options.headers `headers` are custom headers to be sent
 * @param {Any} options.data `data` is the data to be sent as the request body
 * @param {Number} options.timeout`timeout` specifies the number of milliseconds before the request times out. If the request takes longer than `timeout`, the request will be aborted.
 * @param {String} options.responseType `responseType` indicates the type of data that the server will respond with, default: json
 * @param {Object} options.agent `agent` define a custom agent to be used when performing http/https requests, respectively, in node.js. This allows options to be added like `keepAlive` that are not enabled by default.
 */
async function request(options = {}) {
  if (!options.url) {
    throw new Error('request param url is required.');
  }
  const method = (options.method || 'GET').toUpperCase();
  const parsedUrl = new URL(options.url);
  let data = options.data;

  if (['GET', 'HEAD'].indexOf(method) > -1) {
    data &&
      Object.keys(data).forEach((key) => {
        parsedUrl.searchParams.append(key, data[key]);
      });
  }

  const protocol = parsedUrl.protocol;
  const timeout = options.timeout || 15000;
  const headers = Object.assign(
    {
      'Content-Type': 'application/json',
      'User-Agent': `lr/${pkg.version}`,
    },
    options.headers || {}
  );
  const contentType = headers['Content-Type'];
  const isStream =
    data &&
    data.readable !== false &&
    typeof data === 'object' &&
    typeof data.pipe === 'function' &&
    typeof data._read === 'function' &&
	  typeof data._readableState === 'object'

  if (
    Object.prototype.toString.call(data) === '[object Object]' &&
    contentType.startsWith('application/x-www-form-urlencoded')
  ) {
    data = qs.stringify(data);
  } else if (Object.prototype.toString.call(data) !== '[object String]' && !isStream) {
    data = JSON.stringify(data);
  }

  const responseType = options.responseType || 'json';
  const agent = options.agent || (protocol === 'http:' ? http.globalAgent : https.globalAgent);

  return new Promise((resolve, reject) => {
    const requestOptions = {
      protocol: protocol,
      host: parsedUrl.hostname,
      port: parsedUrl.port,
      path: parsedUrl.pathname + parsedUrl.search,
      method: method,
      headers: headers,
      timeout: timeout,
      agent: agent,
    };

    const resHandler = (res) => {
      const statusCode = res.statusCode;

      let response = new Response({
        headers: res.headers,
        statusCode,
        statusMessage: res.statusMessage,
      });

      // 如果 responseType 为 stream 直接返回
      if (options.responseType === 'stream') {
        response.data = res;
        resolve(response);
        return;
      }

      res.on('data', (chunk) => {
        response.addChunk(chunk);
      });
      res.on('end', () => {
        if (responseType === 'json' && statusCode === 200) {
          response.data = response.json();
        } else if (responseType === 'buffer' && statusCode === 200) {
          response.data = response.buffer();
        } else {
          response.data = response.text();
        }
        resolve(response);
      });
      res.on('error', (err) => {
        reject(enhanceError(err, options, null, req));
      });
    };

    const errHandler = (err) => {
      if (req.aborted) {
        return;
      }
      reject(enhanceError(err, options, null, req));
    }

    let req;
    if (protocol === 'http:') {
      req = http.request(requestOptions, resHandler);
    } else if (protocol === 'https:') {
      req = https.request(requestOptions, resHandler);
    } else {
      throw new Error(`Protocol "${protocol}" not supported. Expected "http:" or "https:"`);
    }

    // Handle errors
    req.on('error', errHandler);

    // Handle request timeout
    if (timeout) {
      req.setTimeout(timeout, () => {
        req.abort();

        if (isStream) {
          data.destroy();
        }

        reject(createError('timeout of ' + timeout + 'ms exceeded', options, 'ECONNABORTED', req));
      });
    }

    if (isStream) {
      return pump(data, req, (err) => err && errHandler(err));
    } else if (data && ['GET', 'HEAD'].indexOf(method) === -1) {
      req.write(data);
    }

    req.end();
  });
}

module.exports = request;
